allow `type` and `type alias`

stoeffel opened the pull request:
allows to use sum- and product-types in examples.

```elm
{-|
    type Animal
        = Dog
        | Cat

    double Dog
    --> (Dog, Dog)
-}
double : a -> (a, a)
double a =
    (a, a)
```


related https://github.com/stoeffel/elm-verify/pull/9 (cc @michaeljones)

stoeffel commented:
@eeue56 yes, totally. I was planning to create a file for each example. I think that will lead to less confusing errors. What do you think?

eeue56 commented:
Maybe if the files are only made if there's a type involved? Otherwise it's increasing the IO considerably (and Elm will compile things slower)

stoeffel commented:
### TODO
- [ ] pass all files to write to js once

stoeffel commented on src/VerifyExamples.elm:
```diff
@@ -60,11 +60,13 @@ update msg model =
             in
-                model ! [ writeFile ( moduleName, generatedTestFileText ) ]
+            ( model
+            , writeFiles generatedTests
```
@eeue56 changed it to send everything to js and write these generatedTests in serial

eeue56 replied:
👍

stoeffel commented on run-tests.sh:
```diff
@@ -1,5 +1,5 @@
 #! /bin/bash -ex
-TEST_COUNT=21
+TEST_COUNT=18
```
Yes! it's finally green again

stoeffel commented:
@eeue56 would love another review if you have time. I think I've addressed all your comments from the other PR and I've cleaned some more things. I still don't really like `Compile.elm`, but I think I'll clean that up in a new PR.

stoeffel commented:
reminder : please squash this PR. commits are useless (😿 )

stoeffel commented on src/VerifyExamples/Ast.elm:
```diff
@@ -75,18 +47,38 @@ isExpectiation x =
-    case x of
-        LocalFunction _ _ ->
-            True
+group : List Ast -> { localFunctions : List Ast, imports : List Ast, types : List Ast }
```
maybe `groupExtras`

stoeffel commented on src/VerifyExamples/Ast.elm:
```diff
@@ -97,40 +89,39 @@ astToString ast =
+groupToAst xs =
+    case xs of
+        (IAst.MaybeExpression x) :: rest ->
+            Assertion (String.join "\n" <| x :: List.map IAst.toString rest)
```
use String.Util.unlines here

stoeffel commented on src/VerifyExamples/IntermediateAst.elm:
```diff
@@ -0,0 +1,175 @@
+    Regex.regex "^\\s{4}(\\w+\\s:\\s.*)"
+
+
+oneOf : List (a -> Maybe b) -> a -> Maybe b
```
create a module

eeue56 commented on .travis.yml:
```diff
@@ -32,7 +32,7 @@ install:
 
 script:
-  - npm test
+  - $TRAVIS_BUILD_DIR/sysconfcpus/bin/sysconfcpus -n 2 npm test
```
using `-n 1` is actually more performant than -n 2 :D

stoeffel replied:
oh, yes. forgot about that

eeue56 commented on src/VerifyExamples/Ast.elm:
```diff
@@ -75,18 +48,38 @@ isExpectiation x =
+groupHelp :
+    List Ast
+    -> { localFunctions : List Ast, imports : List Ast, types : List Ast }
+    -> { localFunctions : List Ast, imports : List Ast, types : List Ast }
```
🎨 why not give a name to this type?

stoeffel replied:
sure

eeue56 commented on src/VerifyExamples/Ast.elm:
```diff
@@ -97,40 +90,49 @@ astToString ast =
+            (x :: List.map IAst.toString rest)
+                |> unlines
+                |> Assertion
+                |> Just
```
This common pattern could be a function

stoeffel replied:
:+1:

eeue56 commented on src/VerifyExamples/Comment.elm:
```diff
@@ -0,0 +1,30 @@
+
+commentRegex : Regex
+commentRegex =
+    Regex.regex "({-[^]*?-})\x0D?\n([^\x0D\n(\\s:)(\\s=))]*)\\s[:=]"
```
😱 this is very hard to read

stoeffel replied:
yes, it's horrible. I will do something....

stoeffel commented on src/VerifyExamples/Comment.elm:
```diff
@@ -0,0 +1,41 @@
+commentRegex =
+    Regex.regex <|
+        String.join ""
+            [ "({-[^]*?-})" -- anything between comments
```
@eeue56 does this help?

zwilias replied:
`String.concat` is a useful little `String.join ""` alternative - less thinking about what's going on here.

stoeffel replied:
d'oh! will fix.

stoeffel commented:
@eeue56 addressed all comments.

eeue56 commented on src/String/Util.elm:
```diff
@@ -0,0 +1,21 @@
+escape =
+    Regex.replace All (regex "\\\\") (\_ -> "\\\\")
+        >> Regex.replace All (regex "\\\"") (\_ -> "\\\"")
+        >> Regex.replace All (regex "\\s\\s+") (\_ -> " ")
```
⚠️ potential runtime errors here

stoeffel replied:
will extract. good catch!

eeue56 commented on src/VerifyExamples/Comment.elm:
```diff
@@ -0,0 +1,41 @@
+            , newline
+            , "([^\\s(" ++ newline ++ ")]*)" -- anything that is not a space or newline
+            , "\\s[:=]" -- until ` :` or ` =`
+            ]
```
👍
